### PR TITLE
hashes: remove `all_zeros` and constify constructors

### PR DESCRIPTION
**Pull request by apoelstra**
I think these changes are both uncontroversial but they have fairly large diffs so I am PRing them together before making more invasive changes.

### CODE REVIEW

**Comment by apoelstra** [comment]
Uggh MSRV limitations on constfns in `impl<T> hmac::Hash<T: Tag>`. Will need to rework.

**Comment by apoelstra** [comment]
Ok, this might be controversial now :P. I found a bug in 1.56.1 that lets me bypass the MSRV limitation.

File: bitcoin/src/blockdata/transaction.rs [hunk 1]
```diff
@@ -50,6 +50,24 @@ hashes::hash_newtype! {
 impl_hashencode!(Txid);
 impl_hashencode!(Wtxid);
 
+impl Txid {
+    /// The "all zeros" TXID.
+    ///
+    /// This is used as the "txid" of the dummy input of a coinbase transaction. It is
+    /// ot a real TXID and should not be used in other contexts.
```
Comment (tcharding) [inline bitcoin/src/blockdata/transaction.rs:57]:
```suggestion
    /// not a real TXID and should not be used in other contexts.
```

**Review by tcharding — COMMENTED** [review]
All good except the docs typo!

**Review by tcharding — APPROVED** [review]
ACK 8702e5f74b3b47a899b9a3834bae676c0cdd9410

**Comment by tcharding** [comment]
FTR I do not understand, nor am I going to bother understading, the `(T,)` thing. Since it builds I believe it.

**Comment by apoelstra** [comment]
Fixed typo.

**Comment by apoelstra** [comment]
CI failure is on master -- looks like cargo-public-api changed something out from under us.

I think we should take the opportunity to figure out how to pin cargo-public-api rather than accepting its changes. (Though I guess we eventually will have to.)

**Comment by tcharding** [comment]
Pinned in https://github.com/rust-bitcoin/rust-bitcoin/pull/2893

File: bitcoin/src/pow.rs [hunk 2]
```diff
@@ -1795,8 +1795,8 @@ mod tests {
         // Block 2015, the only information used are `bits` and `time`
         let current = Header {
             version: Version::ONE,
-            prev_blockhash: BlockHash::all_zeros(),
-            merkle_root: TxMerkleNode::all_zeros(),
+            prev_blockhash: BlockHash::from_byte_array([0; 32]),
```
Comment (Kixunil) [inline bitcoin/src/pow.rs:1798]: `BlockHash` still does have `all_zeros` though, why not call it here?

**Review by Kixunil — APPROVED** [review]
ACK b8a21a38de1b54bd3d5ebf175b83a0731ad57565

This is great! I find the hack acceptable because it was stabilized anyway and we want to release 1.0 at higher MSRV anyway.

File: bitcoin/src/pow.rs (hunk 2 above)
Comment (apoelstra) [reply bitcoin/src/pow.rs:1798]: Oh, oops, I must've originally decided not to have `BlockHash::all_zeros` and then reconsidered.

I'll rebase to fix this, sice I want to rebase on the fixed CI anyway.

**Comment by apoelstra** [comment]
Rebased.

**Review by tcharding — APPROVED** [review]
ACK 4652ce20ed8e397d297d0a2e91b61730dcbdf498

**Comment by apoelstra** [comment]
Thanks! Gonna go ahead and merge because the change was nearly trivial (affected only unit tests) from what Kix already ACKed.

**Pull request merged**